File: app/lib/api/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:minister_shared/models/account.dart';
import 'package:minister_shared/models/transaction.dart';
import 'package:minister_shared/models/analytics.dart';
import 'package:minister_shared/models/category_rule.dart';
import '../config.dart';

class ApiClient {
  final _client = http.Client();

  Uri _uri(String path, [Map<String, String>? queryParams]) {
    return Uri.parse('$apiBaseUrl$path').replace(
      queryParameters: queryParams?.isNotEmpty == true ? queryParams : null,
    );
  }

  // Accounts
  Future<List<LinkedAccount>> getAccounts() async {
    final response = await _client.get(_uri('/api/accounts'));
    final list = jsonDecode(response.body) as List<dynamic>;
    return list
        .map((e) => LinkedAccount.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  // Transactions
  Future<TransactionPage> getTransactions({
    String? account,
    String? category,
    String? startDate,
    String? endDate,
    String? search,
    String? sort,
    int page = 1,
    int limit = 50,
  }) async {
    final params = <String, String>{
      'page': page.toString(),
      'limit': limit.toString(),
    };
    if (account != null) params['account'] = account;
    if (category != null) params['category'] = category;
    if (startDate != null) params['startDate'] = startDate;
    if (endDate != null) params['endDate'] = endDate;
    if (search != null) params['search'] = search;
    if (sort != null) params['sort'] = sort;

    final response = await _client.get(_uri('/api/transactions', params));
    return TransactionPage.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<CleanTransaction> updateTransactionCategory(
    String id,
    String category,
  ) async {
    final response = await _client.patch(
      _uri('/api/transactions/$id'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'category': category}),
    );
    final json = jsonDecode(response.body) as Map<String, dynamic>;
    return CleanTransaction.fromJson(
      json['transaction'] as Map<String, dynamic>,
    );
  }

  // Sync
  Future<Map<String, dynamic>> sync([List<String>? accountIds]) async {
    final response = await _client.post(
      _uri('/api/sync'),
      headers: {'Content-Type': 'application/json'},
      body: accountIds != null ? jsonEncode({'accountIds': accountIds}) : null,
    );
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> syncFetch([List<String>? accountIds]) async {
    final response = await _client.post(
      _uri('/api/sync/fetch'),
      headers: {'Content-Type': 'application/json'},
      body: accountIds != null ? jsonEncode({'accountIds': accountIds}) : null,
    );
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> syncClean() async {
    final response = await _client.post(_uri('/api/sync/clean'));
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  // Analytics
  Future<List<CategoryBreakdown>> getCategoryBreakdown({
    String? startDate,
    String? endDate,
    String? account,
    String? category,
  }) async {
    final params = <String, String>{};
    if (startDate != null) params['startDate'] = startDate;
    if (endDate != null) params['endDate'] = endDate;
    if (account != null) params['account'] = account;
    if (category != null) params['category'] = category;

    final response = await _client.get(
      _uri('/api/analytics/categories', params),
    );
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => CategoryBreakdown.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<MonthlyBreakdown>> getMonthlyBreakdown({
    String? startDate,
    String? endDate,
    String? account,
    String? category,
  }) async {
    final params = <String, String>{};
    if (startDate != null) params['startDate'] = startDate;
    if (endDate != null) params['endDate'] = endDate;
    if (account != null) params['account'] = account;
    if (category != null) params['category'] = category;

    final response = await _client.get(_uri('/api/analytics/monthly', params));
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => MonthlyBreakdown.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<WeeklyBreakdown>> getWeeklyBreakdown({
    String? startDate,
    String? endDate,
    String? account,
    String? category,
  }) async {
    final params = <String, String>{};
    if (startDate != null) params['startDate'] = startDate;
    if (endDate != null) params['endDate'] = endDate;
    if (account != null) params['account'] = account;
    if (category != null) params['category'] = category;

    final response = await _client.get(_uri('/api/analytics/weekly', params));
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => WeeklyBreakdown.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  // Category Management
  Future<List<CategoryRule>> getCategoryRules() async {
    final response = await _client.get(_uri('/api/categories'));
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => CategoryRule.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<CategoryRule> createCategoryRule({
    required String category,
    required String pattern,
    bool caseSensitive = false,
  }) async {
    final response = await _client.post(
      _uri('/api/categories'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'category': category,
        'pattern': pattern,
        'caseSensitive': caseSensitive,
      }),
    );

    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to create rule');
    }

    return CategoryRule.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<CategoryRule> updateCategoryRule({
    required String id,
    String? category,
    String? pattern,
    bool? caseSensitive,
  }) async {
    final body = <String, dynamic>{};
    if (category != null) body['category'] = category;
    if (pattern != null) body['pattern'] = pattern;
    if (caseSensitive != null) body['caseSensitive'] = caseSensitive;

    final response = await _client.put(
      _uri('/api/categories/$id'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(body),
    );

    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to update rule');
    }

    return CategoryRule.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<void> deleteCategoryRule(String id) async {
    final response = await _client.delete(_uri('/api/categories/$id'));

    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to delete rule');
    }
  }

  Future<Map<String, dynamic>> shouldImportDefaults() async {
    final response = await _client.get(_uri('/api/categories/should-import'));
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> importDefaultRules() async {
    final response = await _client.post(_uri('/api/categories/import-defaults'));

    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to import defaults');
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<List<CleanTransaction>> getUncategorizedTransactions() async {
    final response = await _client.get(_uri('/api/transactions/uncategorized'));
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => CleanTransaction.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  // Balances
  Future<List<Map<String, dynamic>>> getBalances() async {
    final response = await _client.get(_uri('/api/balances'));
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => e as Map<String, dynamic>)
        .toList();
  }

  Future<List<Map<String, dynamic>>> refreshBalances([
    List<String>? accountIds,
  ]) async {
    final response = await _client.post(
      _uri('/api/balances/refresh'),
      headers: {'Content-Type': 'application/json'},
      body: accountIds != null ? jsonEncode({'accountIds': accountIds}) : '{}',
    );
    return (jsonDecode(response.body) as List<dynamic>)
        .map((e) => e as Map<String, dynamic>)
        .toList();
  }

  Future<void> categorizeTransaction({
    required String id,
    required String category,
    bool createRule = false,
    String? rulePattern,
  }) async {
    await _client.post(
      _uri('/api/transactions/$id/categorize'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'category': category,
        'createRule': createRule,
        'rulePattern': rulePattern,
      }),
    );
  }

  // Settings
  Future<Map<String, dynamic>> resetCategoryRules() async {
    final response = await _client.post(_uri('/api/settings/reset-category-rules'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to reset rules');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> clearCategoryRules() async {
    final response = await _client.post(_uri('/api/settings/clear-category-rules'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to clear rules');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> recategorizeTransactions() async {
    final response = await _client.post(_uri('/api/settings/recategorize'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to recategorize');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> clearCategoryOverrides() async {
    final response = await _client.post(_uri('/api/settings/clear-overrides'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to clear overrides');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> clearTransactions() async {
    final response = await _client.post(_uri('/api/settings/clear-transactions'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to clear transactions');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> unlinkAccounts() async {
    final response = await _client.post(_uri('/api/settings/unlink-accounts'));
    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to unlink accounts');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getStats() async {
    final response = await _client.get(_uri('/api/settings/stats'));
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  String getBackupUrl() {
    return '$apiBaseUrl/api/settings/backup';
  }

  Future<Map<String, dynamic>> restoreBackup(List<int> zipBytes) async {
    final response = await _client.post(
      _uri('/api/settings/restore'),
      headers: {'Content-Type': 'application/zip'},
      body: zipBytes,
    );

    if (response.statusCode >= 400) {
      final errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['error'] ?? 'Failed to restore backup');
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }
}

File: app/src/api/apiClient.ts
import { AccountData } from '../models/account';
import { AppStats, BalanceEntry, CategoryBreakdown, MonthlyBreakdown, WeeklyBreakdown } from '../models/analytics';
import { CategoryRule } from '../models/categoryRule';
import { CleanTransaction, TransactionFilters, TransactionPage } from '../models/transaction';

const BASE_URL = 'http://localhost:3000';

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const url = `${BASE_URL}${path}`;
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    ...options,
  });
  if (res.status >= 400) {
    let msg = `HTTP ${res.status}`;
    try {
      const body = await res.json();
      msg = body.error ?? body.message ?? msg;
    } catch {}
    throw new Error(msg);
  }
  if (res.status === 204) return undefined as T;
  return res.json() as Promise<T>;
}

// ── Accounts ─────────────────────────────────────────────────────────────────

export async function getAccounts(): Promise<AccountData> {
  return request<AccountData>('/api/accounts');
}

// ── Transactions ──────────────────────────────────────────────────────────────

export async function getTransactions(filters: Partial<TransactionFilters> = {}): Promise<TransactionPage> {
  const params = new URLSearchParams();
  if (filters.account) params.set('account', filters.account);
  if (filters.category) params.set('category', filters.category);
  if (filters.startDate) params.set('startDate', filters.startDate);
  if (filters.endDate) params.set('endDate', filters.endDate);
  if (filters.search) params.set('search', filters.search);
  if (filters.sort) params.set('sort', filters.sort);
  if (filters.page != null) params.set('page', String(filters.page));
  const qs = params.toString();
  return request<TransactionPage>(`/api/transactions${qs ? `?${qs}` : ''}`);
}

export async function getUncategorizedTransactions(): Promise<CleanTransaction[]> {
  return request<CleanTransaction[]>('/api/transactions/uncategorized');
}

export async function updateTransaction(
  id: string,
  patch: { category?: string; pinned?: boolean },
): Promise<void> {
  return request<void>(`/api/transactions/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
  });
}

export async function categorizeTransaction(
  id: string,
  category: string,
  createRule: boolean,
): Promise<void> {
  return request<void>(`/api/transactions/${id}/categorize`, {
    method: 'POST',
    body: JSON.stringify({ category, createRule }),
  });
}

// ── Analytics ─────────────────────────────────────────────────────────────────

export interface AnalyticsParams {
  startDate?: string;
  endDate?: string;
  account?: string;
}

export async function getCategoryBreakdown(params: AnalyticsParams = {}): Promise<CategoryBreakdown[]> {
  const qs = new URLSearchParams(params as Record<string, string>).toString();
  return request<CategoryBreakdown[]>(`/api/analytics/categories${qs ? `?${qs}` : ''}`);
}

export async function getMonthlyBreakdown(params: AnalyticsParams = {}): Promise<MonthlyBreakdown[]> {
  const qs = new URLSearchParams(params as Record<string, string>).toString();
  return request<MonthlyBreakdown[]>(`/api/analytics/monthly${qs ? `?${qs}` : ''}`);
}

export async function getWeeklyBreakdown(params: AnalyticsParams = {}): Promise<WeeklyBreakdown[]> {
  const qs = new URLSearchParams(params as Record<string, string>).toString();
  return request<WeeklyBreakdown[]>(`/api/analytics/weekly${qs ? `?${qs}` : ''}`);
}

// ── Category Rules ────────────────────────────────────────────────────────────

export async function getCategoryRules(): Promise<CategoryRule[]> {
  return request<CategoryRule[]>('/api/categories');
}

export async function createCategoryRule(rule: Omit<CategoryRule, 'id'>): Promise<CategoryRule> {
  return request<CategoryRule>('/api/categories', {
    method: 'POST',
    body: JSON.stringify(rule),
  });
}

export async function updateCategoryRule(id: string, rule: Partial<Omit<CategoryRule, 'id'>>): Promise<CategoryRule> {
  return request<CategoryRule>(`/api/categories/${id}`, {
    method: 'PUT',
    body: JSON.stringify(rule),
  });
}

export async function deleteCategoryRule(id: string): Promise<void> {
  return request<void>(`/api/categories/${id}`, { method: 'DELETE' });
}

export async function importDefaultRules(): Promise<void> {
  return request<void>('/api/categories/import-defaults', { method: 'POST' });
}

export async function shouldImportDefaults(): Promise<boolean> {
  const res = await request<{ shouldImport: boolean }>('/api/categories/should-import-defaults');
  return res.shouldImport;
}

// ── Sync ──────────────────────────────────────────────────────────────────────

export async function syncAll(): Promise<void> {
  return request<void>('/api/sync', { method: 'POST' });
}

export async function syncFetch(): Promise<void> {
  return request<void>('/api/sync/fetch', { method: 'POST' });
}

export async function syncClean(): Promise<void> {
  return request<void>('/api/sync/clean', { method: 'POST' });
}

// ── Balances ──────────────────────────────────────────────────────────────────

export async function getBalances(): Promise<BalanceEntry[]> {
  return request<BalanceEntry[]>('/api/balances');
}

export async function refreshBalances(accountIds?: string[]): Promise<void> {
  return request<void>('/api/balances/refresh', {
    method: 'POST',
    body: JSON.stringify({ accountIds }),
  });
}

// ── Settings / Admin ──────────────────────────────────────────────────────────

export async function getStats(): Promise<AppStats> {
  return request<AppStats>('/api/stats');
}

export async function resetCategoryRules(): Promise<void> {
  return request<void>('/api/settings/reset-categories', { method: 'POST' });
}

export async function clearCategoryRules(): Promise<void> {
  return request<void>('/api/settings/clear-categories', { method: 'POST' });
}

export async function recategorize(): Promise<void> {
  return request<void>('/api/settings/recategorize', { method: 'POST' });
}

export async function clearOverrides(): Promise<void> {
  return request<void>('/api/settings/clear-overrides', { method: 'POST' });
}

export async function clearPins(): Promise<void> {
  return request<void>('/api/settings/clear-pins', { method: 'POST' });
}

export async function clearTransactions(): Promise<void> {
  return request<void>('/api/settings/clear-transactions', { method: 'POST' });
}

export async function unlinkAccounts(): Promise<void> {
  return request<void>('/api/settings/unlink-accounts', { method: 'POST' });
}

export function getBackupUrl(): string {
  return `${BASE_URL}/api/backup`;
}

export async function restoreBackup(fileUri: string): Promise<void> {
  const formData = new FormData();
  formData.append('file', {
    uri: fileUri,
    name: 'backup.zip',
    type: 'application/zip',
  } as unknown as Blob);
  return request<void>('/api/backup/restore', {
    method: 'POST',
    headers: { 'Content-Type': 'multipart/form-data' },
    body: formData as unknown as BodyInit,
  });
}
